### DISCUSSION TITLE
new quizzes

### DISCUSSION THREAD

**Discussion by vincentvmarshburn** [ucfopen/canvasapi#628] [Q&A]
I was quite intrigued by the release of REST endpoints for new quizzes in Canvas, and was especially interested to see that the latest version of canvasapi was incorporating this functionality. I have been testing out the methods and arguments for a new quiz object, and have encountered a complication that my meager brain cannot seem to overcome.

So far, many of the settings work as expected. I am able to readily get a new quiz object with:

````
newQuiz = course.get_new_quiz(assignmentID)
````

Then, I am able to modify the new quiz settings by passing such arguments as:

```
examSettings = {'quiz': {'title': 'New Quiz v11'}}
examSettings = {'quiz': {'unlock_at':dateStart.isoformat(), 'lock_at': dateEnd.isoformat()}}
examSettings = {'quiz': {'quiz_settings': {'has_time_limit': True, 'session_time_limit_in_seconds': maxTime}}}
examSettings = {'quiz': {'quiz_settings': {'require_student_access_code': True, 'student_access_code': 'Slartibartfast'}}}
```

using the "update" method:
````
update = newQuiz.update(**examSettings)
````

However, I am experiencing some difficulties with the IP filter settings for the new quiz. I can enable or disable the IP filter setting well enough with this:
````
examSettings = {'quiz': {'quiz_settings': {'filter_ip_address': True}}}
examSettings = {'quiz': {'quiz_settings': {'filter_ip_address': False}}}
````

But I cannot for the life of me figure out how to post the list of IP address ranges. Based on my understanding of the Canvas documentation, it seems it is supposed to be in the form of a list of lists:
````
examSettings = {'quiz': {'quiz_settings': {'filters': {'ips': [['192.168.1.0', '192.168.1.255'], ['172.16.0.0', '172.16.255.255']]}}}}
````

When I attempt to post this using the "update" method for the new quiz, I receive a "status code 502" error.

If I manually apply the IP filter to the new quiz using the Web interface and then view the "quiz_settings" property of the new quiz object, the format of the IP range seems to coincide with that described above:
````
{... 'filters': {'ips': [['192.168.1.0', '192.168.1.255'], ['172.16.0.0', '172.16.255.255']]} ...}
````

I have attempted all sorts of variations for the format of the IP address range, but to no avail. I am wondering if there is something simple I am overlooking. Any insight or advice would be greatly appreciated.

**Comment by jamiehomewood** [comment]
Did you have to do anything special to get this working?  I cant for the life of me get the get_new_quiz call to work!  I always get the 'Course' object has no attribute 'get_new_quiz'

**Reply by Thetwam** [reply]
@jamiehomewood What version of CanvasAPI are you running? Run `pip show canvasapi` to see. New Quizzes support was added with version 3.2.0. If you're running an older version, run `pip install --upgrade canvasapi` to get the latest code.

**Reply by jamiehomewood** [reply]
@Thetwam using 3.2.0 and still doesn't want to work.  Can a package be installed in different places?

**Reply by vincentvmarshburn** [reply]
As far as I can determine, all I have done is update to 3.2.0, as Thetwam described above. My virtual environment has loads of modules, but I have not updated any of them specifically in relation to canvasapi. This is what I get when i run `pip show canvasapi`:

```
Name: canvasapi
Version: 3.2.0
Summary: API wrapper for the Canvas LMS
Home-page: https://github.com/ucfopen/canvasapi
Author: University of Central Florida - Center for Distributed Learning
Author-email: techrangers@ucf.edu
License: MIT License
Location: c:\users\vmarshbu\downloads\winpython-64bit-3.6.3.0qt5\virtualenv1113\lib\site-packages
Requires: arrow, pytz, requests
Required-by:
```
Regarding the indicated dependencies, it looks like I have arrow 1.2.3, pytz 2021.3, and requests 2.26.0.

**Reply by bennettscience** [reply]
I'm assuming you did, but make sure your virtual environment is active when you do `pip install --upgrade canvasaspi`. It may be installed globally but not locally in your project.

You could also try a one-liner like this to see which version the interpreter is actually using: 

```python
python -c "import canvasapi; print(canvasapi.__version__)"
```

**Reply by jamiehomewood** [reply]
Thanks all, sorted it now.  It was how I installed Python way back when and was using a different location!  I was thinking that but couldn't figure out where.

**Comment by Thetwam** [comment]
I suspect this is misbehaving due to how CanvasAPI handles lists in complex arguments. Ultimately, we take the keyword arguments that are passed into the function and convert them into form-data. I spent a little bit of time in Postman trying different formats and can't seem to find a version it likes. Either it returns a 500, 502, or when it does 200 the values are unchanged. If someone can figure out the valid version in form-data to send, that'd be very helpful.

We could also look into sending the request as JSON instead of form-data.
